datafusion/src/physical_plan/cross_join.rs

```diff
@@ -0,0 +1,315 @@
+                children[0].clone(),
+                children[1].clone(),
+            )?)),
+            _ => Err(DataFusionError::Internal(
+                "CrossJoinExec wrong number of children".to_string(),
+            )),
+        }
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        self.right.output_partitioning()
+    }
+
+    async fn execute(&self, partition: usize) -> Result<SendableRecordBatchStream> {
+        // we only want to compute the build side once
+        let left_data = {
+            let mut build_side = self.build_side.lock().await;
+
+            match build_side.as_ref() {
+                Some(stream) => stream.clone(),
+                None => {
+                    let start = Instant::now();
+
+                    // merge all left parts into a single stream
+                    let merge = MergeExec::new(self.left.clone());
+                    let stream = merge.execute(0).await?;
+
+                    // Load all batches and count the rows
+                    let (batches, num_rows) = stream
+                        .try_fold((Vec::new(), 0usize), |mut acc, batch| async {
+                            acc.1 += batch.num_rows();
+                            acc.0.push(batch);
+                            Ok(acc)
+                        })
+                        .await?;
+                    let merged_batch =
+                        concat_batches(&self.left.schema(), &batches, num_rows)?;
+                    *build_side = Some(merged_batch.clone());
+
+                    debug!(
+                        "Built build-side of cross join containing {} rows in {} ms",
+                        num_rows,
+                        start.elapsed().as_millis()
+                    );
+
+                    merged_batch
+                }
+            }
+        };
+
+        let stream = self.right.execute(partition).await?;
+
+        if left_data.num_rows() == 0 {
+            return Ok(Box::pin(MemoryStream::try_new(
+                vec![],
+                self.schema.clone(),
+                None,
+            )?));
+        }
+
+        Ok(Box::pin(CrossJoinStream {
+            schema: self.schema.clone(),
+            left_data,
+            right: stream,
+            right_batch: Arc::new(std::sync::Mutex::new(None)),
+            left_index: 0,
+            num_input_batches: 0,
+            num_input_rows: 0,
+            num_output_batches: 0,
+            num_output_rows: 0,
+            join_time: 0,
+        }))
+    }
+}
+
+/// A stream that issues [RecordBatch]es as they arrive from the right  of the join.
+struct CrossJoinStream {
+    /// Input schema
+    schema: Arc<Schema>,
+    /// data from the left side
+    left_data: JoinLeftData,
+    /// right
+    right: SendableRecordBatchStream,
+    /// Current value on the left
+    left_index: usize,
+    /// Current batch being processed from the right side
+    right_batch: Arc<std::sync::Mutex<Option<RecordBatch>>>,
```

Comment:
I wonder if we should be using a `tokio::sync::Mutex` or a `futures::Mutex` here instead of `std::sync::Mutex` which could block the tokio executor thread

Reply:
Tokio says about it https://tokio-rs.github.io/tokio/doc/tokio/sync/struct.Mutex.html

`The primary use case of the async mutex is to provide shared mutable access to IO resources such as a database connection. If the data stored behind the mutex is just data, it is often better to use a blocking mutex such as the one in the standard library or parking_lot`

Maybe we can switch (had some difficulty using it with async inside this method)?

What do you think?

Reply:
I think that while the `std::sync::Mutex` is held, it will tie up a tokio executor thread (can't execute other futures) so if it is held for any length of time, we should use `tokio::sync::Mutex` which will allow the executor thread to run other futures

Reply:
I believe the lock is in scope in both uses for a very short amount of time as no actual processing is done while having the lock (just `clone().unwrap()` or mutating the `right_batch`) so I think the above documentation applies for this "just data" usage, no? 

I made the first use a bit more clear to show that the lock is only held for `unwrapping`.

Reply:
I think you are correct @Dandandan  -- I just re-reviewed the code and I am not sure why I thought that the left batch was not buffered or that the lock was a problem

I agree that using a `std::sync::Mutex` to `clone()` a record batch is just fine -- I think I had originally read this code as holding that lock while the right hand side was being streamed in.

Sorry for the confusion